File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> m = new HashMap<>();
        int j= 0;
        int i=0;
        int n = s.length();
        int res = 0;
        while(j< n){
            int c = m.getOrDefault(s.charAt(j), 0);
            m.put(s.charAt(j), c+1);
            while(i<=j && m.size() < j-i+1){
                c = m.get(s.charAt(i));
                if(c==1)
                    m.remove(s.charAt(i));
                else
                    m.put(s.charAt(i), c-1);
                i++;
            }
            res = Math.max(res, j-i+1); 
            j++;
        }
        return res;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        int n = s.length();
        unordered_map<char, int> m;
        int i=0;
        for(int j=0;j<n;j++){
            if(m.find(s[j]) == m.end()){
                m[s[j]] =1;
            }
            else{
                m[s[j]] += 1;
            }
            while(i < j && j-i+1 > m.size()){
                m[s[i]] -= 1;
                if(m[s[i]]==0)
                    m.erase(s[i]);
                i++;
            }
            res = max(res, j-i+1);
        }
        return res;
    }
};